### PR TITLE
Starlight Polish (Space is blue!)

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Adds support to underlays to realize_overlays
Ensures decals properly handle plane offsets
Fixes space lighting double applying if it's changeturf'd into. this will be important later
Makes solar vis_contents block emissives as expected
Moves transit tube overlays to update_overlays, adds emissive blockers to them

#### Adds render steps

An expansion on render_target based emissive blockers. 
They allow us to hijack an object's appearance and draw it somewhere else, or even modify it, THEN draw it somewhere else. 
They chain quite nicely

Fixes shuttles deleting z holder objects

#### Makes space emissive, makes walls and floors block emissives
The core idea here goes like this:
We make space glow, and give its overlays some color

This way, the tile and space parallax remain fullbright, along with anything that doesn't block emissives, but anything that does block emissives will instead get shaded the color of starlight

This requires a bit of extra work, see later

This is done automatically with render relays, which now support specifiying layer and color (Need to make an editor for these one of these days)

The emissive blocking floor stuff requires making a second render plate to prevent double scaling

Also adds some new layering defines for lighting, and ensures all turf lights have a layer. We'll get to this soon

#### Makes things in space blue

We color them the same as starlight, by taking advantage of space being emissive
This means that things in space that block emissive will block it correctly and be colored blue by the light overlay, but space itself
will remain fullbright

This does require redefining what always_lit means, but nothing but cordons use that so it's fineee

#### Makes glass above space glow, and some other stuff

Glass tiles that sit above space will now shine light with matching
color to the glasses color. This includes mat tiles.

Glass tiles (not mat because they have no alpha) also only partially block emissives.
Adds a new proc that uses render steps to acomplish this, essentially we're cutting out bits below X alpha and drawing what remains as an emissive.

#### Modifies partial space showing to support glow

Essentially, alongside displaying space as an underlay, we also display a light overlay colored like starlight.
That starlight overlay gets masked to only be visible in bits that do not contain any alpha.

We also mask the turf lighting to not go into bits that have no alpha, to ensure we get the effect we want.
This is done with that lighting layer thing I mentioned earlier.

#### Makes appearance realization's list output ordered

I want it output in order of overlay, sub overlay suboverlay, next overlay
Need to use insert for that

## Why It's Good For The Game

Pretty!
Also having space be emissive is a very very good way to test for fucked emissive blockers (If it's broken why are we even drawing the overlay)
I know for a fact mob blockers on lizards and socks are kinda yorked, I think there's more

<details>
<summary>
Old
</summary>

image
image
image

</details>

<details>
<summary>
New
</summary>

image
image
image

</details>

## Changelog

:cl:
add: Space now makes things in it starlight faintly blue
fix: Glass floors that display space now properly let space shine through them, rather then hiding it in the dark
add: Glass floors above space now glow faintly depending on their glass type
/:cl:

### CODE REVIEW

**Comment by Elli-Skala** [comment]
Will breaches / fake space tiles also have the blue effect in station?

**Comment by CRITAWAKETS** [comment]
good, but maybe a little too blue

**Comment by calliophis-maculiceps** [comment]
> good, but maybe a little too blue

yeah could be a little darker

File: code/__HELPERS/lighting.dm [hunk 1]
```diff
@@ -31,3 +31,31 @@
 	var/mutable_appearance/appearance = mutable_appearance(icon, icon_state, layer, offset_spokesman, EMISSIVE_PLANE, alpha, appearance_flags | EMISSIVE_APPEARANCE_FLAGS, offset_const)
 	appearance.color = GLOB.em_block_color
 	return appearance
+
+/// Takes a non area atom and a threshold
+/// Makes it block emissive with any pixels with more alpha then that threshold, with the rest allowing the light to pass
+/// Returns a list of objects, automatically added to your vis_contents, that apply this effect
+/// QDEl them when appropriate
```
Comment (ZephyrTFA) [inline code/__HELPERS/lighting.dm:38]:
```suggestion
/// QDEL them when appropriate
```

File: code/modules/visuals/render_steps.dm [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	layer = FLOAT_LAYER
+	mouse_opacity = MOUSE_OPACITY_TRANSPARENT
+	//Why?
+	//render_targets copy the transform of the target as well, but vis_contents also applies the transform
+	//we'll display using that, so we gotta reset
+	appearance_flags = KEEP_APART|KEEP_TOGETHER|RESET_TRANSFORM
+
+/atom/movable/render_step/Initialize(mapload, source)
+	. = ..()
+	verbs.Cut() //Cargo cultttttt
```
Comment (ZephyrTFA) [inline code/modules/visuals/render_steps.dm:17]: Why is this needed?

**Comment by ArthuroPep** [comment]
I agree with the others, this is good but it is a little too blue, make it a light cyan tinge or something

**Comment by Wallemations** [comment]
> > good, but maybe a little too blue
> 
> yeah could be a little darker

I think the opposite, I think the blue is too dark as it is and is making it seem like space isn't as bright as it should be.
Pulling back on it by either moving it more towards white or lowering the opacity of the effect would likely be good

**Comment by LemonInTheDark** [comment]
> Will breaches / fake space tiles also have the blue effect in station?

yes

File: code/modules/visuals/render_steps.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/modules/visuals/render_steps.dm:17]: cargo cultttttt from old lighting code

**Comment by Inari-Whitebear** [comment]
Make starlight colour depend on which star orbit the station gets!

**Comment by Jacquerel** [comment]
I think “black orbit” and such are just codewords like “red alert” rather than a literal description.

**Comment by Inari-Whitebear** [comment]
> I think “black orbit” and such are just codewords like “red alert” rather than a literal description.

Sad, would be neat

**Comment by AdipemDragon** [comment]
> **Makes things in space blue**
> We color them the same as starlight, by taking advantage of space being emissive

I know that, between the title and example images, it's all based on the one single color, but. Would it be way too hard to actually "color them the same as starlight"? After all:
image
Space is not always blue.

(Or at the least, would this pr allow for an easy followup where it's actually set to glow the space's color?)

**Comment by LemonInTheDark** [comment]
They are colored the same as starlight, what you're seeing is gas clouds that spawn at random.
In theory yes it would be possible, would just need to pull picking that out into some other process outside of parallax, and make glass blend its color with the color of the starlight somehow

File: code/__DEFINES/layers.dm [hunk 3]
```diff
@@ -47,9 +47,15 @@
...
+#define EMISSIVE_PLANE 13
 /// This plane masks out lighting to create an "emissive" effect, ie for glowing lights in otherwise dark areas.
-#define EMISSIVE_PLANE 14
+#define EMISSIVE_RENDER_PLATE 14
 #define EMISSIVE_RENDER_TARGET "*EMISSIVE_PLANE"
+// Ensures all the render targets that point at the emissive plate layer correctly
+#define EMISSIVE_Z_BELOW_LAYER 1
+#define EMISSIVE_FLOOR_LAYER 2
+#define EMISSIVE_SPACE_LAYER 3
+#define EMISSIVE_WALL_LAYER 4
```
Comment (Time-Green) [inline code/__DEFINES/layers.dm:58]: Could you autodoc the ones you added 😃

File: code/game/turfs/open/floor/glass.dm [hunk 4]
```diff
@@ -15,6 +15,11 @@
 	heavyfootstep = FOOTSTEP_GENERIC_HEAVY
 	floor_tile = /obj/item/stack/tile/glass
 	overfloor_placed = FALSE
+	var/list/glow_stuff = list()
```
Comment (Time-Green) [inline code/game/turfs/open/floor/glass.dm:18]: autodoc

**Review by Time-Green — CHANGES_REQUESTED** [review]
Looks fine and I haven't seen any big complaints yet

File: code/__DEFINES/layers.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/__DEFINES/layers.dm:58]: it would just be THIS LAYER ISUSED FOR THE Z BELOW
THIS LAYER IS USED FOR THE FLOOR

can do if you'd like but I think it'd suck

**Comment by LemonInTheDark** [comment]
Hold on a sec this looks slightly off with multiz and I don't wanna snooker you, lemme grab a photo

**Comment by LemonInTheDark** [comment]
image
It double layers on multiz, which produces semi unwanted effects.
I could solve the stuff in space doing this, but I don't have a good way of handling the on turf stuff. guess it is kinda a vibe tho

**Comment by Time-Green** [comment]
> image It double layers on multiz, which produces semi unwanted effects. I could solve the stuff in space doing this, but I don't have a good way of handling the on turf stuff. guess it is kinda a vibe tho

It looks pretty cool though, I approve

**Comment by Profakos** [comment]
I hope the aurora will be finally visible now...
